src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs

```diff
@@ -563,18 +563,28 @@ protected override void Generate(RenameTableOperation operation, [CanBeNull] IMo
                 .Append(IndexColumnList(operation.Columns, operators))
                 .Append(")");
 
-            if (!string.IsNullOrEmpty(operation.Filter))
-            {
-                builder
-                    .Append(" WHERE ")
-                    .Append(operation.Filter);
-            }
+            IndexOptions(operation, model, builder);
 
             builder.AppendLine(';');
 
             EndStatement(builder);
         }
 
+        protected override void IndexOptions(CreateIndexOperation operation, IModel model, MigrationCommandListBuilder builder)
+        {
+            var includeProperties = (string[])operation[NpgsqlAnnotationNames.IndexInclude];
+
+            if (includeProperties != null && includeProperties.Length > 0)
```

Comment:
It seems like we should do a `VersionAtLeast` check here. I'm thinking of a situation where I want to test the same model across PG10 and PG11, but without a compatibility check, I'm stuck commenting out `ForNpgsqlInclude` each time.

Reply:
I can see what you mean, but I'm not sure if I agree (yet!)... Here's what I'm thinking:

_I have the server version stored in `appsettings.json`. Because the translation code uses `VersionAtLeast`, I don't have to use conditional flags to control server/client decisions about where functions execute. When I upgrade PosgreSQL, I just update `appsettings.json` and things keep working, maybe with new functions being translated. All of this means that I don't need to recheck my query code._

_If migrations also make use of `VersionAtLeast`, then I also don't need to use conditional flags when configuring my model. I can use `ForNpgsqlInclude`, and it won't have an effect unless the configuration file has a sufficient version specified._

At this point, I think this is the type of feature versioning that is best to handle within the provider, since we have clear visibility into its impact (e.g. works with 11.0+, breaks if used below). 

At the same time, I think it's a fair argument that having too much magic in the migration feature would be a bad thing. Though here, I think we're still in safe territory, since this would follow similar uses of `VersionAtLeast` (e.g. handling identity columns).

What do you think?

Reply:
The idea of playing around with PostgreSQL versions on the same context/application seems pretty dangerous to me... For example, let's not forget that migrations frequently aren't executed, but rather exported to SQL scripts which are later executed as part of deployment. These scripts are of course tied to a specific version of PostgreSQL.

Another point is that whether a covering index exists or not is quite a crucial performance aspect of an application - I would not expect the application to run as-is with or without it. I'm trying to say that if I have a context using covering indices on PG 11, and someone comes along and says "oh let's run this on PG 10", I think we *don't* want the application to silently ignore the covering indices: the application performance can potentially degrade in a dramatic way. I'd rather it produced errors, and that the user were forced to manually remove them (or set up an explicit condition if they really want to be multi-version).

In other words, I'm not sure it should be a goal to transparently switch between versions and have features magically work differently, hiding the potential differences... As you say, I'm trying to avoid introducing too much magic...

Let me know if this convinces you :) If not I'll be happy to hear what you think, and we can also turn to @YohDeadfall for his opinion.

Reply:
> For example, let's not forget that migrations frequently aren't executed, but rather exported to SQL scripts which are later executed as part of deployment. These scripts are of course tied to a specific version of PostgreSQL.

> I think we _don't_ want the application to silently ignore the covering indices: the application performance can potentially degrade in a dramatic way. I'd rather it produced errors, and that the user were forced to manually remove them (or set up an explicit condition if they really want to be multi-version).

Those are great points. I was thinking about these as if they were all in the same bucket, but you've highlighted a clear distinction between this type of versioning and versioning for server/client translation. Consider me convinced!